qiskit/optimization/problems/optimization_problem.py

```diff
@@ -342,227 +358,470 @@ def set_problem_type(self, _type):
+    variables"""
+
+    CONST = -1
+
+    def __init__(self):
+        self._src: OptimizationProblem = None
+        self._dst: OptimizationProblem = None
+        self._subs = {}
+
+    def substitute_variables(self, src: OptimizationProblem,
+                             constants: Optional[SparsePair] = None,
+                             variables: Optional[SparseTriple] = None):
+        """Substitutes variables with constants or other variables.
+
+        Args:
+            src: an optimization problem whose variables will be substituted
+            constants: replace variable by constant
+                i.e., SparsePair.ind (variable) -> SparsePair.val (constant)
+
+            variables: replace variables by weighted other variable
+                need to copy everything using name reference to make sure that indices are matched
+                correctly
+                i.e., SparseTriple.ind1 (variable)
+                        -> SparseTriple.ind2 (variable) * SparseTriple.val (constant)
+
+        Returns:
+            tuple[OptimizationProblem, SubstitutionStatus]:
+                If the resulting problem has no issue, the status is `success`.
+                Otherwise, an empty problem and status `infeasible` are returned.
+
+        Raises:
+            QiskitOptimizationError: if the substitution is invalid as follows.
+                - Same variable is substituted multiple times.
+                - Coefficient of variable substitution is zero.
         """
 
+        self._src = src
+        self._dst = OptimizationProblem()
+        self._dst.set_problem_name(src.get_problem_name())
+        # do not set problem type, then it detects its type automatically
+
+        self._subs_dict(constants, variables)
+
+        results = [
+            self._variables(),
+            self._objective(),
+            self._linear_constraints(),
+            self._quadratic_constraints(),
+        ]
+        if any(r == SubstitutionStatus.infeasible for r in results):
+            ret = SubstitutionStatus.infeasible
+        else:
+            ret = SubstitutionStatus.success
+        return self._dst, ret
+
+    @staticmethod
+    def _feasible(sense: str, rhs: float, range_value: float = 0) -> bool:
+        """Checks feasibility of the following condition
+            0 `sense` rhs
+        """
+        # pylint: disable=misplaced-comparison-constant
+        if sense == 'E':
+            if 0 == rhs:
+                return True
+        elif sense == 'L':
+            if 0 <= rhs:
+                return True
+        elif sense == 'G':
+            if 0 >= rhs:
+                return True
+        else:  # sense == 'R'
+            if range_value >= 0:
+                if rhs <= 0 <= rhs + range_value:
+                    return True
+            else:
+                if rhs + range_value <= 0 <= rhs:
+                    return True
+        return False
+
+    @staticmethod
+    def _dict_to_list(op, dic):
```

Comment:
The name `_dict_to_list` and the fact that it is a staticmethod suggests that this does something very basic, like `list = list(dict.keys())` or so. This does something a little more elaborate and also requires an optimization problem.
Could you rename this to something more descriptive? Like `_to_index_and_values` or `_unroll_function_definition` or so?